src/pvt.c

```diff
@@ -278,6 +278,163 @@ static u8 filter_solution(gnss_solution* soln, dops_t* dops)
+    if ((update = pvt_solve(rx_state, n_used, nav_meas, omp, H)) > 0) {
+      break;
+    }
+  }
+
+  if (iters >= PVT_MAX_ITERATIONS) {
+    /* Reset state if solution fails */
+    rx_state[0] = 0;
+    rx_state[1] = 0;
+    rx_state[2] = 0;
+    return -1;
+  }
+
+  return 0;
+}
+
+/* Return values:
+ *    1: repaired solution, using one fewer observation
+ *       returns prn of removed measurement if removed_prn ptr is passed
+ *   -1: no reasonable solution possible
+ *
+ *  Results stored in rx_state, omp, H
+ */
+static s8 pvt_repair(double rx_state[],
+                     const u8 n_used,
+                     const navigation_measurement_t nav_meas[n_used],
+                     double omp[n_used],
+                     double H[4][4],
+                     u8 *removed_prn)
+{
+  /* Try solving with n-1 navigation measurements. */
+  navigation_measurement_t nav_meas_copy[n_used];
+  u8 one_less = n_used - 1;
+  s8 bad_sat = -1;
+  u8 num_passing = 0;
+
+  memcpy(nav_meas_copy, nav_meas, n_used * sizeof(navigation_measurement_t));
+
+  /* Carefully ordered */
+  for (s8 drop = one_less; drop >= 0; drop--) {
+    /* Swaps the last omitted value with the one at index `drop'.
+     * On first iteration, does nothing (omits last nav_meas) */
+    navigation_measurement_t temp;
+    temp = nav_meas_copy[drop];
+    nav_meas_copy[drop] = nav_meas_copy[one_less];
+    nav_meas_copy[one_less] = temp;
+
+    s8 flag = pvt_iter(rx_state, n_used - 1, nav_meas_copy, omp, H);
+    if (flag == -1) {
+      /* Didn't converge. */
+      continue;
+    }
+
+    if (residual_test(n_used-1, omp, rx_state, 0)) {
+      num_passing++;
+      bad_sat = drop;
+    }
+  }
+
+  if (num_passing == 1) {
+    /* Repair is possible by omitting bad_sat. Recalculate that solution. */
+    memcpy(nav_meas_copy, nav_meas, n_used * sizeof(navigation_measurement_t));
+    nav_meas_copy[bad_sat] = nav_meas_copy[one_less];
+    s8 flag = pvt_iter(rx_state, n_used - 1, nav_meas_copy, omp, H);
+    assert(flag == 0);
+    if (removed_prn) {
+      *removed_prn = nav_meas[bad_sat].prn;
+    }
+    return 1;
+  } else {
+    return -1;
+  }
+}
+
+/* Return values:
+ *    0: inital solution ok
+ *    1: repaired solution, using one fewer observation
+ *       returns prn of removed measurement if removed_prn ptr is passed
+ *   -1: no reasonable solution possible
+ *
+ *  Results stored in rx_state, H
+ */
+static s8 pvt_solve_and_check(double rx_state[],
+                              const u8 n_used,
+                              const navigation_measurement_t nav_meas[n_used],
+                              double H[4][4],
+                              u8 *removed_prn)
+{
+  double omp[n_used];
+  double residual = 0;
+
+  if (n_used < 4) {
+    return -1;
+  }
+
+  s8 flag = pvt_iter(rx_state, n_used, nav_meas, omp, H);
+
+  if (flag == 0 && residual_test(n_used, omp, rx_state, &residual)) {
+    /* Solution ok. */
+    return 0;
```

Comment:
We can pass this state out to the user in the solution message so they can decide not to trust the solution if RAIM is not available.